### DISCUSSION TITLE
should task runner execute linux bash shell with login (-l)?

### DISCUSSION THREAD

**Discussion by amiramw** [eclipse-theia/theia#8818] [Q&A] [answered]
Theia task runne runs bash shell task with `-l` which means that bash act as if it had been invoked as a login shell. It means that current `PATH` is ignored and then calculated using `.profile`.

https://github.com/eclipse-theia/theia/blob/f1e37fbdbe9cf0bdcd30f58d2c8fbf6d6808419e/packages/task/src/node/process/process-task-runner.ts#L178-L181

I believe (but not 100% sure) that in vscode `-l` is only used sometimes for Mac. For bash it seems that only `-c` is used:

https://github.com/microsoft/vscode/blob/10d92e5efaf67a39660260f4fa5e44b3f847c597/src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts#L47-L65
https://github.com/microsoft/vscode/blob/f13720627f4720528d7642775c0f6e272ef637f1/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L1040-L1059

Any reason we execute bash with `-l` or can we just remove it?

@marechal-p

**Comment by paul-marechal** [comment]
I think I was fighting with shells oddities and this flag might have been a desperate attempt at reading user's configurations files to correctly expand files from his path. Thank you for looking into this, I read https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Bash-Startup-Files into more details, and doing my own tests it looks like you are right indeed. `-l` should not be used in this case.

**Reply by paul-marechal** [reply]
Although, should we point `BASH_ENV` to the current user's home `.bashrc` file? Without this, I think we are missing aliases and other things configured through `.bashrc` (e.g. aliases; e.g. `nvm`, ...)

**Reply by paul-marechal** [reply]
Did some more experiments:

- Simply pointing `BASH_ENV` to `~/.bashrc` does not fix issues like invoking `nvm` because we find the following script at the top:

  ```bash
  # If not running interactively, don't do anything
  case $- in
      *i*) ;;
        *) return;;
  esac
  ```

  This prevents aliases and such from being loaded when invoking bash as `bash -c "..."`.
- `nvm` tries to install itself in `~/.bash_profile` first, and only installs itself under `~/.bashrc` as a last resort.
  See: https://github.com/nvm-sh/nvm#install--update-script.
- This makes me guess that we would preferably want to use the `~/.bash_profile` mechanisms and that it would be the right thing to try and run `bash -l -c "..."`, but if fails on most setup because people don't properly setup their `~/.bash_profile` configuration file? (I sure didn't)

**Reply by paul-marechal** [reply]
If `-l` causes more trouble than what it tries to solve, I am fine getting rid of it.

But I tried to explore our options, and I still don't know what's the best to do at the end of the day :)

**Reply by amiramw** [reply]
what was the issue with nvm that `-l` solved?

**Reply by paul-marechal** [reply]
If the `nvm` function is defined in `.bash_profile` then adding `-l` makes it work. If `.bash_profile` is also written to source `.bash_aliases` then `-l` makes the aliases available.

**Reply by amiramw** [reply]
and do you know if the same flow in vscode also has this issue? and then maybe we don't need to handle it.

**Reply by paul-marechal** [reply]
Well, if you want to align with VS Code they have the same issue. Go for it then.

**Reply by paul-marechal** [reply]
@amiramw Actually I don't understand your issue. How is the path corrupted by loading `.profile`? What is the content of that file?

**Reply by amiramw** [reply]
I'm running theia on docker image based on debian. In the docker file I set `PATH` via `ENV` statement.
The `.profile` I get is probably pretty standard:
```sh
# ~/.profile: executed by the command interpreter for login shells.
# This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login
# exists.
# see /usr/share/doc/bash/examples/startup-files for examples.
# the files are located in the bash-doc package.

# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.
#umask 022

# if running bash
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
	. "$HOME/.bashrc"
    fi
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin:$PATH"
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/.local/bin" ] ; then
    PATH="$HOME/.local/bin:$PATH"
fi
```

Anyway, if I run `bash -l` the PATH variable will always be `/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games` and ignore what I put in the docker file:
```sh
user@d16913c66d12:~$ bash -l
user@d16913c66d12:~$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

I can solve this also by aligning PATHs between my docker file and .profile file, but if we want to align tasks with vscode then maybe we shouldn't run without `-l`.

BTW, I confirmed that vscode on linux (suse) runs a shell task on bash without changing the PATH so it probably runs without `-l` as I see in the code.

**Reply by paul-marechal** [reply]
I see nothing in your `~/.profile` that would completely replace the previous `PATH`. When trying a debian image I saw that `/etc/profile` actually resets the `PATH` to what you just reported, so I am able to reproduce indeed. I would love to use user configuration files, but given this issue it seems reasonable to remove `-l` when invoking bash on Linux. Thanks for helping me understand your problem.

**Comment by vince-fugnitto** [comment, marked as answer]
I believe it is answered with https://github.com/eclipse-theia/theia/pull/8834 being merged.
